### PR TITLE
add new property of Application model for oauth2.0 grant type

### PR DESCRIPTION
**Pull request by vinebrancho**
OAuth2 server internally use `Application` model for registering its oauth2 clients. But `Application` model doesn't have any property to set permitted grant type for registred client. And currently  loopback-component-oauth2 doesn't check grant type any where when it handles request of token creation from clients. For checking grant type on request of access token, I think that `Application` model has grant type property as first step, and then as second step, loopback-component-oauth2 should check if grant type of requesting client (clientId/clientSecret of http header) is same to value of grant type set to `Application` model.

For the first step, I simply prepare this patch

### CODE REVIEW

**Comment by slnode** [comment]
Can one of the admins verify this patch? To accept patch and trigger a build add comment ".*ok\W+to\W+test.*"

File: common/models/application.json [hunk 1]
```diff
@@ -30,6 +30,10 @@
     "email": "string",
     "emailVerified": "boolean",
 
+    "grantType": {
+      "type": "string",
+      "description": "Grant type permitted to the application for OAuth 2.0"
+    },
```
Comment (raymondfeng) [inline common/models/application.json]: The name should be `grantTypes` and the type should be `["string"]`.

Comment (vinebrancho) [inline common/models/application.json]: I changed type of `grantType` to array and updated my patch.

File: common/models/application.json [hunk 2]
```diff
@@ -30,6 +30,10 @@
     "email": "string",
     "emailVerified": "boolean",
 
+    "grantType": {
```
Comment (raymondfeng) [inline common/models/application.json]: The name should be `grantTypes`

Comment (vinebrancho) [inline common/models/application.json]: Ok. New patch for that is uploaded

**Comment by raymondfeng** [comment]
This is going to be breaking change. For example, if you already have a database table in MySQL, an `autoupdate` is needed to add the new column. I wonder what's the best way to merge/release this. @bajtos @ritch any ideas?

**Comment by bajtos** [comment]
Can we provide two models: base `Application` and then `OAuth2Application`? Or perhaps implement the new property as a mixin?

IMO this is a good example why the component-specific per-Application configuration should not be a part of a single god-like model Application. We should have per-component models instead and setup a relation Application hasOne (embedsOne) ComponentConfig. This is true for both OAuth2 and Push.

I suppose mixins are a good compromise how to keep the component-specific properties inside components while providing a single Application object for the consumers.

**Comment by vinebrancho** [comment]
Is there any update regarding this?

**Comment by OwenBrotherwood** [comment]
Just a comment: the PR is failing, npm test shows:

```

  1) loopback loopback.configureModel(ModelCtor, config) updates existing settin
gs:
     AssertionError: expected { Object (strict, ttl, ...) } to not have property
 'base'
      at Context.<anonymous> (c:\Users\ob_000\vinebrancho\loopback\test\loopback
.test.js:360:42)
      at callFn (c:\Users\ob_000\vinebrancho\loopback\node_modules\grunt-mocha-t
est\node_modules\mocha\lib\runnable.js:223:21)
      at Test.Runnable.run (c:\Users\ob_000\vinebrancho\loopback\node_modules\gr
unt-mocha-test\node_modules\mocha\lib\runnable.js:216:7)
      at Runner.runTest (c:\Users\ob_000\vinebrancho\loopback\node_modules\grunt
-mocha-test\node_modules\mocha\lib\runner.js:373:10)
      at c:\Users\ob_000\vinebrancho\loopback\node_modules\grunt-mocha-test\node
_modules\mocha\lib\runner.js:451:12
      at next (c:\Users\ob_000\vinebrancho\loopback\node_modules\grunt-mocha-tes
t\node_modules\mocha\lib\runner.js:298:14)
      at c:\Users\ob_000\vinebrancho\loopback\node_modules\grunt-mocha-test\node
_modules\mocha\lib\runner.js:308:7
      at next (c:\Users\ob_000\vinebrancho\loopback\node_modules\grunt-mocha-tes
t\node_modules\mocha\lib\runner.js:246:23)
      at Immediate.<anonymous> (c:\Users\ob_000\vinebrancho\loopback\node_module
s\grunt-mocha-test\node_modules\mocha\lib\runner.js:275:5)
      at Immediate._onImmediate (c:\Users\ob_000\vinebrancho\loopback\node_modul
es\continuation-local-storage\node_modules\async-listener\glue.js:188:31)

  2) User User.create Email is required:
     Uncaught AssertionError: [ 'presence', 'format.null' ] deepEqual [ 'presenc
e', 'format.blank' ]
      at c:\Users\ob_000\vinebrancho\loopback\test\user.test.js:80:16
      at c:\Users\ob_000\vinebrancho\loopback\node_modules\loopback-datasource-j
uggler\lib\dao.js:264:9
      at ModelConstructor.<anonymous> (c:\Users\ob_000\vinebrancho\loopback\node
_modules\loopback-datasource-juggler\lib\validations.js:484:13)
      at ModelConstructor.next (c:\Users\ob_000\vinebrancho\loopback\node_module
s\loopback-datasource-juggler\lib\hooks.js:75:12)
      at done (c:\Users\ob_000\vinebrancho\loopback\node_modules\loopback-dataso
urce-juggler\lib\validations.js:481:25)
      at c:\Users\ob_000\vinebrancho\loopback\node_modules\loopback-datasource-j
uggler\lib\validations.js:555:7
      at c:\Users\ob_000\vinebrancho\loopback\node_modules\continuation-local-st
orage\node_modules\async-listener\glue.js:188:31
      at process._tickDomainCallback [as _tickCallback] (node.js:381:11)

  3) User User.login should handle multiple `include`:
     Uncaught AssertionError: body.user: expected undefined to not equal undefin
ed
      at Test.<anonymous> (c:\Users\ob_000\vinebrancho\loopback\test\user.test.j
s:344:50)
      at Test.assert (c:\Users\ob_000\vinebrancho\loopback\node_modules\supertes
t\lib\test.js:221:6)
      at Server.assert (c:\Users\ob_000\vinebrancho\loopback\node_modules\supert
est\lib\test.js:132:12)
      at Server.g (events.js:199:16)
      at Server.emit (events.js:104:17)
      at net.js:1392:10
      at c:\Users\ob_000\vinebrancho\loopback\node_modules\continuation-local-st
orage\node_modules\async-listener\glue.js:188:31
      at process._tickDomainCallback [as _tickCallback] (node.js:381:11)


```

**Comment by bajtos** [comment]
Hi @vinebrancho, sorry for keeping this pull request stale for so long. I re-read the comments and also checked your proposed changes. I think the easiest way how to make this happen is to implement the new property directly in loopback-component-oauth2.  
Perhaps in [lib/models/index.js](https://github.com/strongloop/loopback-component-oauth2/blob/4e70cc2e5f7a4e8f4d2ff4708c1ce5071fa78a09/lib/models/index.js#L29-L52) where we are already modifying application's models.

I am closing this pull request as rejected.

**Pull request closed without merging**